pkg/hubble/getters.go

```diff
@@ -0,0 +1,165 @@
+	ep := getter.manager.LookupIP(ip)
+	if ep == nil {
+		return nil, false
+	}
+	return hubbleEndpoint.ParseEndpointFromModel(ep.GetModel()), true
+}
+
+// GetNamesOf implements DNSGetter.GetNamesOf. It looks up DNS names of a given IP from the
+// FQDN cache of an endpoint specified by sourceEpID.
+func (getter *LocalEndpointDNSGetter) GetNamesOf(sourceEpID uint64, ip net.IP) []string {
+	ep := getter.manager.LookupCiliumID(uint16(sourceEpID))
+	if ep == nil {
+		return nil
+	}
+	return ep.DNSHistory.LookupIP(ip)
+}
+
+// LocalIPGetter implements IPGetter interface from Hubble.
+//  - IPGetter: https://github.com/cilium/hubble/blob/04ab72591faca62a305ce0715108876167182e04/pkg/parser/getters/getters.go#L46
+type LocalIPGetter struct {
+	ipCache *ipcache.IPCache
+}
+
+// NewLocalIPGetter returns an initialized pointer to LocalIPGetter.
+func NewLocalIPGetter(ipCache *ipcache.IPCache) *LocalIPGetter {
+	return &LocalIPGetter{ipCache: ipCache}
+}
+
+// GetIPIdentity returns the IP identity of the given IP address. Hubble uses this function to populate
+// fields like namespace and pod name for remote endpoints. If the K8s metadata is unavailable, it sets
+// the Identity field for the IP identity.
+func (getter *LocalIPGetter) GetIPIdentity(ip net.IP) (hubbleIPCache.IPIdentity, bool) {
+	ipIdentity, ok := getter.ipCache.LookupByIP(ip.String())
+	if !ok {
+		return hubbleIPCache.IPIdentity{}, false
+	}
+	meta := getter.ipCache.GetK8sMetadata(ip.String())
+	if meta == nil {
+		return hubbleIPCache.IPIdentity{
+			Identity: ipIdentity.ID,
+		}, true
+	}
+	return hubbleIPCache.IPIdentity{
+		Identity:  ipIdentity.ID,
+		Namespace: meta.Namespace,
+		PodName:   meta.PodName,
+	}, true
+}
+
+// LocalServiceGetter implements the ServiceGetter interface from Hubble.
+//  - ServiceGetter: https://github.com/cilium/hubble/blob/04ab72591faca62a305ce0715108876167182e04/pkg/parser/getters/getters.go#L52
+type LocalServiceGetter struct {
+	svc *service.Service
+}
+
+// NewLocalServiceGetter returns an initialized pointer to LocalServiceGetter.
+func NewLocalServiceGetter(svc *service.Service) *LocalServiceGetter {
+	return &LocalServiceGetter{svc: svc}
+}
+
+// GetServiceByAddr looks up service by IP/port. Hubble uses this function to annotate flows
+// with service information.
+func (g *LocalServiceGetter) GetServiceByAddr(ip net.IP, port uint16) (hubbleProto.Service, bool) {
+	addr := loadbalancer.L3n4Addr{
+		IP: ip,
+		L4Addr: loadbalancer.L4Addr{
+			Port: port,
+		},
+	}
+	svc, ok := g.svc.GetDeepCopyServiceByAddr(addr)
+	if !ok {
+		return hubbleProto.Service{}, false
+	}
+	return hubbleProto.Service{
+		Name:      svc.Name,
+		Namespace: svc.Namespace,
+	}, true
+}
+
+// GetServiceByID looks up service by ID.
+func (g *LocalServiceGetter) GetServiceByID(id int64) (hubbleProto.Service, bool) {
+	svc, ok := g.svc.GetDeepCopyServiceByID(loadbalancer.ServiceID(id))
```

Comment:
This is extremely costly. Why are we performing deepcopies if we just want the name and namespace of a service? Can't we simplify this function to be less CPU / memory intensive?

Reply:
this function was removed from the getter interface, so i deleted it.